### PR TITLE
postgresql: refactor and cleanup

### PR DESCRIPTION
**Pull request by wolfgangwalther**
## Description of changes

TLDR: This is a pure refactor, where `postgresql/default.nix` is split into multiple files and some `passthru` attributes are cleaned up. No derivations are changed.

Goal: I am currently working on building PG v16 via meson instead of autotools. This will make it possible to make `pkgsStatic.postgresql` work - at least for building `libpq`, which should cover almost all use-cases. This PR does not implement the meson build, yet. But to be able to have autotools-based builds up to v15 and meson-based builds for v16+ separated without repeating all the shared parts, this PR moves things around a little bit and splits them up in multiple files. But even without building with meson, the new structure is still an improvement, I think.

This is best reviewed commit-by-commit. I tried to make it reviewable despite the big chunks of code moved around. The most important fact: The derivations for all 5 versions with and without `_jit` don't change, so it's a pure refactor.

The files are structured like this:
- `default.nix` contains the logic to create all the top-level attributes, both with and without `_jit` suffix.
- `<major>.nix` files contain the version-specific stuff: version number, hashes, musl patches. This is recommended [here](https://github.com/NixOS/nixpkgs/tree/master/pkgs#versioning).
- `generic.nix` has all the metadata and most passthru arguments of the derivation, thus dealing with all the packages/extensions, tests and jit-helpers. It also contains all the patches, to make sure that both autotools and meson based builds will stay in sync.
- `autotools.nix` only has those parts which are specific to the autotools builds. This makes it easy to add a `meson.nix` next to it later.

Along the way, I also cleaned up a few things around arguments and passthru attributes:
- There was a `readline` passthru attribute introduced around 20 years ago - but entirely unused. Removed.
- `psqlSchema` is auto-created instead of passed for each version. Less chance for error when copy&pasting for a new version.
- `jitSupport` / `llvm` are not passed as passthru attributes to extensions, but via `callPackage` / the new pkgs scope. Again, reduces the chance for mistakes.
- The `this` argument is removed and replaced with internal recursion. This makes it possible to actually have things like `(postgresql.override { jitSupport = true; }).pkgs.postgis` work correctly without making sure to pass `this` all the time.

I also have commits to clean up `thisAttr` which suffers from the same problem for tests that `this` did - but those would not be a pure refactor anymore, so I'll put them in a separate PR.

Pinging maintainers: @thoughtpolice @danbst @globin @marsam @ivan @Ma27 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Ma27** [comment]
May take until next week to get to it. But I'm definitely interested in reviewing that, thanks!

**Comment by Ma27** [comment]
Hmm, try running `nix-instantiate -A postgresql --argstr system x86_64-darwin`? Or whatever attribute is affected.

You don't need such hardware for the evaluation itself, that only gets relevant when it comes to building stuff, but the ofborg error looks like it didn't even get to that.

**Comment by Ma27** [comment]
Oh sorry I only saw the backtrace and didn't scroll down entirely :sweat_smile:

**Review by wolfgangwalther — COMMENTED** [review]
Note: The latest push was just a rebase on latest master, no changes.

File: pkgs/servers/sql/postgresql/generic.nix [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    disallowedReferences = [ stdenv'.cc ];
+
+    passthru = let
+      this = self.callPackage generic args;
+      jitToggle = this.override {
+        jitSupport = !jitSupport;
+      };
+    in
+    {
+      psqlSchema = lib.versions.major version;
```
Comment (Ma27) [inline pkgs/servers/sql/postgresql/generic.nix:224]: I'm wondering if we should still expose `jitSupport` here, that's the easiest way for other packages to check if postgresql is built with or without that.

File: nixos/modules/services/databases/postgresql.md [hunk 2]
```diff
@@ -277,7 +277,7 @@ self: super: {
 Here's a recipe on how to override a particular plugin through an overlay:
 ```
 self: super: {
-  postgresql_15 = super.postgresql_15.override { this = self.postgresql_15; } // {
+  postgresql_15 = super.postgresql_15.override {} // {
```
Comment (Ma27) [inline nixos/modules/services/databases/postgresql.md:280]: The `.override {}` shouldn't be needed anymore here.

File: pkgs/servers/sql/postgresql/generic.nix (hunk 1 above)
Comment (Ma27) [reply pkgs/servers/sql/postgresql/generic.nix:224]: So, the way I thought about this was: we effectively have two variants of postgresql packaged, JIT and non-JIT (and multiple releases, but those can be found out via `.version`) and I think it doesn't hurt to expose the (rather prominent) detail of the package.

One thing I could imagine is a reusable module that expects to have a JIT variant for $service and asserts against that. It's not the reponsibility of that module to set the postresql package if it's for a different service and I don't think you want that because you'll run into issues quickly then when you'll need to patch the package on your own.

File: pkgs/servers/sql/postgresql/generic.nix [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+      ./patches/specify_pkglibdir_at_runtime.patch
+      ./patches/findstring.patch
+
+      (substituteAll {
+        src = ./patches/locale-binary-path.patch;
+        locale = "${if stdenv.isDarwin then darwin.adv_cmds else lib.getBin stdenv.cc.libc}/bin/locale";
+      })
+
+    ] ++ lib.optionals stdenv'.hostPlatform.isMusl (
+      map fetchurl (lib.attrValues muslPatches)
```
Comment (Ma27) [inline pkgs/servers/sql/postgresql/generic.nix:124]: I'm wondering why this was `fetchurl` in the first place, maybe we should switch back to `fetchpatch`. You should check if the hashes are different then.
If so, this needs to go to a follow-up PR.

**Comment by Ma27** [comment]
Hmm... would you mind trying out if adding 2da789a409c67c99c7e7b27730256d42f10a5601 to `.git-blame-ignore-revs` helps to retain a bit of the history of the previous file?

No idea if that works out, but maybe worth a try?

Anyways, I like that change and the fact that we got a few cleanups ready without any rebuilds.

Will let a few other people review as well before merging. cc @marsam @thoughtpolice @ivan

**Comment by Ma27** [comment]
> I think that's a great idea, not sure whether to wait to do this in a second PR (maybe in the follow up cleanup PR)

Fair.

File: pkgs/servers/sql/postgresql/generic.nix (hunk 1 above)
Comment (Ma27) [reply pkgs/servers/sql/postgresql/generic.nix:224]: > But JIT support doesn't change PostgreSQL functionally, afaik

To my knowledge, yes. But modules warning in case of unsuitable configurations is still a legit use-case (and I can certainly imagine applications where you want that).

> Personally, I could see more benefit in exposing enableSystemd or enablePython

I think you're right.
Postgresql extensions have a nice way to discover that and if you really need it you can still use `hasInfix "-jit-"`.

The core problem behind this is that we just don't have a good configuration story for packages (as in, you can read from `config.services.foo` how service `foo` is configured, but you can't do the same for `pkgs.foo` out of the box), so this should be solved at a different level.

> btw, just noting: Should we maybe rename jitSupport to enableJIT for consistency

Yes, but I'd suggest to leave `jitSupport ? null` around for 1 release and raise a warning if it's not null (i.e. `true`/`false`). `enableJIT` should be set to `jitSupport` in that case, then. Sure, NixOS users won't even notice because they'll probably use the module, but I'm pretty sure there are a few nixpkgs users out there.

These package arguments not only have horrible discovery, but as a result also horrible (i.e. non-existant) ways to communicate changes (short of "This attribute isn't supposed to be part of this function argument").

**Comment by Ma27** [comment]
> To make this work, we will have to split the commit into two commits

Hmm, there's really no other way to trick git into detecting the change otherwise? :/ 

I mean, we're trading the ability to bisect postgresql here.
That said, I think it should be possible to fix the issue with the other rename by `callPackage`ing `postgresql/generic.nix` in the first commit, right?

**Comment by Ma27** [comment]
> I think this is in good shape

Agreed.

Ping @ajs124 @marsam @yu-re-ka you were the last folks who touched the package itself.

I'd consider merging it at some point next week though, I'd rather not stall this.

File: pkgs/servers/sql/postgresql/generic.nix (hunk 3 above)
Comment (marsam) [reply pkgs/servers/sql/postgresql/generic.nix:127]: `fetchurl` was used because those patches are static files from AlpineLinux. On the other hand, `fetchpatch` is recommended when the patch is generated dynamically https://github.com/NixOS/nixpkgs/issues/240141
Nonetheless, I'm not against changing it to `fetchpatch`

**Review by marsam — APPROVED** [review]
LGTM. Good work, thank you!

**Review by Ma27 — APPROVED** [review]
Great work!
